runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	db := sqlx.NewDb(otelsql.OpenDB(connector), "duckdb")
+
+	err = otelsql.RegisterDBStatsMetrics(db.DB, otelsql.WithAttributes(attribute.String("db.system", "duckdb"), attribute.String("db_identifier", d.dbIdentifier)))
+	if err != nil {
+		return nil, fmt.Errorf("registering db stats metrics: %w", err)
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	err = d.attachDBs(ctx, db, path, read)
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	// 2023-12-11: Hail mary for solving this issue: https://github.com/duckdblabs/rilldata/issues/6.
+	// Forces DuckDB to create catalog entries for the information schema up front (they are normally created lazily).
+	// Can be removed if the issue persists.
+	_, err = db.ExecContext(context.Background(), `
+		select
+			coalesce(t.table_catalog, current_database()) as "database",
+			t.table_schema as "schema",
+			t.table_name as "name",
+			t.table_type as "type", 
+			array_agg(c.column_name order by c.ordinal_position) as "column_names",
+			array_agg(c.data_type order by c.ordinal_position) as "column_types",
+			array_agg(c.is_nullable = 'YES' order by c.ordinal_position) as "column_nullable"
+		from information_schema.tables t
+		join information_schema.columns c on t.table_schema = c.table_schema and t.table_name = c.table_name
+		group by 1, 2, 3, 4
+		order by 1, 2, 3, 4
+	`)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+func (d *db) attachDBs(ctx context.Context, db *sqlx.DB, path string, read bool) error {
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	var views []string
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		// NOTE :: we always look at the write version
+		// Tables in read path are removed after getting a new handle
+		// So we need to always look at the write version to ensure we do not reattach dropped tables
+		version, exist, _ := d.writeTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		versionPath := filepath.Join(path, entry.Name(), version)
+
+		// read meta file
+		f, err := os.ReadFile(filepath.Join(versionPath, "meta.json"))
+		if err != nil {
+			_ = os.RemoveAll(versionPath)
+			d.logger.Warn("error in reading meta file", slog.String("table", entry.Name()), slog.Any("error", err))
+			return err
+		}
+		var meta meta
+		err = json.Unmarshal(f, &meta)
+		if err != nil {
+			_ = os.RemoveAll(versionPath)
+			d.logger.Warn("error in unmarshalling meta file", slog.String("table", entry.Name()), slog.Any("error", err))
+			return err
+		}
+
+		if meta.ViewSQL != "" {
+			// table is a view
+			views = append(views, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s\n)", safeSQLName(entry.Name()), meta.ViewSQL))
+			continue
+		}
+		switch meta.Format {
+		case BackupFormatDB:
+			dbName := dbName(entry.Name())
+			var readMode string
+			if read {
+				readMode = " (READ_ONLY)"
+			}
```

Comment:
Even in write mode, shouldn't it mount all the dependency tables as readonly? Once a table version is committed, it will never be mutated, right?

Reply:
As of now the reads and writes are separate so I took some liberties in `Alter` and `Insert` like directly updating the db and then renaming the directory to avoid copy. If we go ahead with non duplicated path then yes tables must not be mutated and we will need to attach in `READ_ONLY` mode.